src/main/scala/org/dia/algorithms/mcc/MCSUtils.scala

```diff
@@ -31,125 +31,155 @@ import org.dia.tensors.AbstractTensor
+    val (srcNode, dstNode) = (edge.srcNode, edge.destNode)
+    val (srcKey, dstKey) = (srcNode.hashKey(), dstNode.hashKey())
+    (MCSNodeMap(srcKey), MCSNodeMap(dstKey))
+  }
 
-    /** Extract the node mask from the MCCNode metadata
-     * @param thisNode MCCNode the current node
-     * @param lats Array[Double] representing the latitudes
-     * @param lons Array[Double] representing the longitudes
-     * @param tightestBox Boolean to use tightest box around data.
-     */
-    def extract_masked_data(thisNode: MCCNode, lats: Array[Double], lons: Array[Double], tightestBox: Boolean): Unit = {
-        var nodeID = ""
-        var latMin = .0
-        var latMax = .0
-        var lonMin = .0
-        var lonMax = .0
-        var lonMinOffset = 0
-        var lonMaxOffset = 0
-        var latMinOffset = 0
-        var latMaxOffset = 0
-        if (tightestBox == true) {
-            latMin = lats.minBy(v => math.abs( v - (thisNode.getLatMin())))
-            latMax = lats.minBy(v => math.abs( v - (thisNode.getLatMax())))
-            lonMin = lons.minBy(v => math.abs( v - (thisNode.getLonMin())))
-            lonMax = lons.minBy(v => math.abs( v - (thisNode.getLonMax())))
-        }
-        else {
-            latMin = lats(0)
-            latMax = lats.last
-            lonMin = lons(0)
-            lonMax = lons.last
-        }
-        latMinOffset = lats.indexOf(latMin)
-        latMaxOffset = lats.indexOf(latMax)
-        lonMinOffset = lons.indexOf(lonMin)
-        lonMaxOffset = lons.indexOf(lonMax)
-        var nodeGrid = new ArrayInt.D2(((latMaxOffset - latMinOffset) + 1), ((lonMaxOffset - lonMinOffset) + 1))
-        var ima = nodeGrid.getIndex()
-        val gridMap: mutable.HashMap[String, Double] = thisNode.grid
-        gridMap.foreach{case(k, v) =>
-            val indices = k.replace("(", "").replace(")", "").replace(" ", "").split(",")
-            nodeGrid.setDouble(ima.set( (indices(0).toInt) - latMinOffset, (indices(1).toInt) - lonMinOffset), v.toInt)
-        }
-        nodeID = "/tmp/F" + thisNode.getFrameNum.toString + "CE" + thisNode.getCloudElemNum.toString.dropRight(2) +
-          "_" + latMinOffset.toString + "_" + latMaxOffset.toString + "_" + lonMinOffset.toString + "_" +
-          lonMaxOffset.toString + ".nc"
-        netcdf_write(nodeID, nodeGrid, lats, lons)
+
+  /** Extract the node mask from the MCCNode metadata
+   *
+   * @param thisNode    MCCNode the current node
+   * @param lats        Array[Double] representing the latitudes
+   * @param lons        Array[Double] representing the longitudes
+   * @param tightestBox Boolean to use tightest box around data.
+   * @return Tuple consisting of the nodeID, and grid array
+   */
+  def extract_masked_data(thisNode: MCCNode,
+                          lats: Array[Double],
+                          lons: Array[Double],
+                          tightestBox: Boolean): (String, ArrayInt.D2) = {
```

Comment:
I wonder if there's a better way to return the ID and Array, instead of using a tuple.